### ISSUE TITLE
The approximate_gradient function seems not follow the equations in paper

### ISSUE THREAD

**Issue by GuanlinLee** [Jianbo-Lab/HSJA#3]
https://github.com/Jianbo-Lab/HSJA/blob/6a9145643f342b62790712ef1917de14276a200a/hsja.py#L180

It seems like you use fval as the average of the subtraction between the Binary function and the baseline of the estimate. I do not understand this. Is this another approximation?

On the other hand, referring to eq.11, for me, it is more like an estimate of integration. Could you give me some details? Thanks.

**Comment by GuanlinLee** [comment]
And there is a difference of gradf between  [https://github.com/Jianbo-Lab/HSJA/blob/master/hsja.py](url) and [https://github.com/bethgelab/foolbox/blob/master/foolbox/attacks/bapp.py](url). It seems like you use a wrong indentation to calculate the gradf.

**Comment by Jianbo-Lab** [comment]
Hi @GuanlinLee There is a slight difference in the implementation of gradf in two versions. But in fact, they are the same thing. This is not a wrong indentation.

See Equation (18) for the introduction of baseline. It is used to reduce variance, as was proved in Theorem 3.

**Comment by GuanlinLee** [comment]
But you just map the output of decision function from {0,1} to {-1,1}. I can not understand why fval is corresponding to the average of the subtraction between the Binary function and the baseline.

**Comment by Jianbo-Lab** [comment]
I didn't quite understand what you mean. Can you clarify what you think I am doing incorrectly, and what you think the right way is? (perhaps in code)

The reason I map the output to {-1,1} is because this is how the response from the model is defined in our work. The reason why we need to subtract the average by the baseline, as I said previously, is because we use Equation (18).

**Comment by GuanlinLee** [comment]
In my opinion, calculating the gradf is ought to

`fval = 2 * decisions.astype(float).reshape(decision_shape) - 1.0`
`fval_avg -= np.mean(fval)`
`fval=fval+fval_avg`
`gradf = np.mean(fval * rv, axis = 0) \\for each elements in fval`

by following my understand of eq.18. First, calculate the mean of fval. Then, subtract it from every element in fval. Finally, multiply u_b by each element in fval and average.

**Comment by Jianbo-Lab** [comment]
So mine is
fval = 2 * decisions.astype(float).reshape(decision_shape) - 1.0
fval_avg = np.mean(fval)
fval=fval-fval_avg
gradf = np.mean(fval * rv, axis = 0) \\for each elements in fval

Our difference is these two lines:
fval_avg -= np.mean(fval)
fval=fval+fval_avg

It seems you used an undefined variable fval_avg. What is that?

**Comment by GuanlinLee** [comment]
I made a mistake about fval-=np.mean(fval). I recognized it as fval=-np.mean(fval) so that we need to subtract it. It is clear now.
But it is still unclear to me how to estimate the gradients vie MC method. And I did not find any reference paper about that. Is eq.11 not an integration?

**Comment by Jianbo-Lab** [comment]
okay, so we have the same understanding at the code level.

The MC estimate of gradient was proposed in Equation 11. It is not an integration. You may take our work as a reference, where Theorem 2 argues why this MC estimate is an asymptotically unbiased one. 

It is not difficult to see that the baseline-included MC estimate shares exactly the same property described in Theorem 2. And its variance is further reduced based on Theorem 3.

**Comment by GuanlinLee** [comment]
Thanks a lot for your help. I think my questions have been solved.

**Comment by machanic** [comment]
@Jianbo-Lab Which gradient does HSJA estimated for? (in function "approximate_gradient")
The following code returns a true gradient but it does not similar to the approximate gradient returned by `def approximate_gradient`
I have posted this question to https://github.com/Jianbo-Lab/HSJA/issues/14
Thanks 
```
def get_grad(self, model, x, true_labels, target_labels):
        with torch.enable_grad():
            # x = torch.clamp(x,min=0,max=1.0).cuda()
            x = x.cuda()
            x.requires_grad_()
            logits = torch.softmax(model(x),dim=1)
            if true_labels is not None:
                true_labels = true_labels.cuda()
            if target_labels is not None:
                target_labels = target_labels.cuda()
            loss = self.cw_loss(logits, true_labels, target_labels)
            gradient = torch.autograd.grad(loss, x,retain_graph=True)[0].cpu().detach()
        return gradient
```
